text/0045-api.md

```diff
@@ -0,0 +1,191 @@
+To be able to call the functions of the`Service` trait, input data needs to be converted.
+A conversion from `JSON` to `TransactionsByBundleParams` could be defined as follows:
+```
+impl TryFrom<&JsonValue> for TransactionsByBundleParams {
+    type Error = &'static str;
+    fn try_from(value: &JsonValue) -> Result<Self, Self::Error> {
+        let entry = match value["entry"].as_str() {
+            Some(str) => HashItem::try_from(str)?.0,
+            None => return Err("can not find entry hash"),
+        };
+        let bundle = match value["bundle"].as_str() {
+            Some(str) => HashItem::try_from(str)?.0,
+            None => return Err("can not find bundle hash"),
+        };
+        Ok(TransactionsByBundleParams { entry, bundle })
+    }
+}
+```
+The same applies to responses - a conversion back from `TransactionsByBundleResponse` to `JSON` could be defined like this:
+```
+impl From<TransactionsByBundleResponse> for JsonValue {
+    fn from(res: TransactionsByBundleResponse) -> Self {
+        let mut json_obj = Map::new();
+        for (hash, tx_ref) in res.tx_refs.iter() {
+            json_obj.insert(
+                String::from(&HashItem(hash.clone())),
+                JsonValue::from(&TransactionRefItem(tx_ref.clone())),
+            );
+        }
+        JsonValue::Object(json_obj)
+    }
+}
+```
+
+Notable is the use of `HashItem` or `TransactionRefItem` in the conversions. These wrapper types are used to be conform with Rust's orphan rules. Since multiple services use for example the `Hash` type within parameters/response, conversions should be reusable.
+For this reason, `TryFrom/From`s should be defined also for inner types (e.g. `Hash`) of a request/response.
+```
+pub struct HashItem(pub Hash);
+
+impl TryFrom<&str> for HashItem {
+    type Error = &'static str;
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        ...
+    }
+}
+
+impl From<&HashItem> for String {
+    fn from(value: &HashItem) -> Self {
+        ...
+    }
+}
+
+impl From<&HashItem> for JsonValue {
+    fn from(value: &HashItem) -> Self {
+        JsonValue::String(String::from(value))
+    }
+}
+```
+
+Some types require more conversion work than others. For consistency - if desired - `TryFrom`/`From`s could be defined for all inner types of a request/response. An example that requires less conversion work, but still might get its `From` implementation:
+```
+pub struct MilestoneIndexItem(pub MilestoneIndex);
+
+impl From<&MilestoneIndexItem> for JsonValue {
+    fn from(value: &MilestoneIndexItem) -> Self {
+        JsonValue::from(*value.0)
+    }
+}
+```
+
+## WebService trait
+
+The `WebService` trait specifies which endpoints a Web service should implement. This also means, it doesn't need to expose all functions of the `Service` trait.
+
+```
+#[async_trait]
+pub trait WebService {
```

Comment:
Extend from above. And maybe here we will have a meta trait like as a common service what this kind of service should have:

```suggestion
trait WebService: QueryService + UserService {
```